### DISCUSSION TITLE
Override Admin Theme Styles Without New Layout File

### DISCUSSION THREAD

**Discussion by fuzl-llc** [OrchardCMS/OrchardCore#12549] [Q&A] [answered]
I want to make some very minor changes to the admin site, currently just changing some css. I've created a theme based on TheAdmin, added my own custom admin.css file to wwwroot/css, added the "admin" tag to the theme project manifest, and selected it as the default admin theme in the admin site. I know I can create my own new Layout file, and reference my new .css file at /MyAdminTheme/css/admin.css. However, I am just fine with the existing layout file and don't want to make any changes except to have my custom css. I even want to keep the existing TheAdmin.css and just have mine override its styles in a few places. Is there a way to just have my styles "injected" after the existing .css and leave everything else as-is?

**Comment by Skrypt** [comment]
I never tried but you could try and create a ResourceManifestOptionsConfiguration.cs file in your inherited admin theme which would do:

```c#
            _manifest
                .DefineStyle("admin")
                .SetUrl("~/TheAdminInherited/css/TheAdmin.min.css", "~/TheAdminInherited/css/TheAdmin.css")
                .SetVersion("1.0.0");
```

This should override the actual "admin" styles assuming that you copied over the /Assets folder and Assets.json file in your own inherited theme to actually generate these assets in your custom admin theme wwwroot folder.

It is possible that the ResourceManifestOptionsConfiguration returns an error if you define 2 styles with the same name. Then, if you get this issue, let us know.

**Reply by ns8482e** [reply]
@fuzl-llc  `ResourceManifestOptionsConfiguration` is a way to add custom css/js without modifying layout view.
In above example,  `TheAdmin.css` is resource defined as `style` you can have your own resources (css/js) can be defined same way.

Once you defined using `ResourceManifestOptionsConfiguration`,  you register it in startup.cs.  
Then you need to render it in your views that depends on your css. In your `cshtml` or `liquid` you can use style/script tag helper to let OC know that you want to render given resource.    

eg..
```cshtml
<style asp-name="my-custom-admin-css" asp-src="~/MyCustomAdminTheme/css/my-custom-admin.min.css" debug-src="~/MyCustomAdminTheme/css/my-custom-admin.css" depends-on="admin"></style>
```


See docs https://docs.orchardcore.net/en/latest/docs/reference/modules/Resources/#inline-definition

**Reply by Skrypt** [reply]
The only thing is that you still need to define a Layout.cshtml file for this to work. So, maybe injecting the element in the header would also work for this with some custom code.

**Reply by fuzl-llc** [reply]
> you still need to define a Layout.cshtml file for this to work

...and I basically have no views in my theme. I just want to override one CSS style right now. One of the built-in editors doesn't have enough margin around it for my site so I just want to increase the margin and do some other minor CSS tweaks to the admin layout. No new view, no new functionality, just very minor CSS additions.

**Reply by fuzl-llc** [reply]
> maybe injecting the element in the header would also work for this

I'm pretty sure this is doable but it seemed weird and not ideal - if there is no other supported or built-in way, then maybe I'll give it a try...

**Comment by fuzl-llc** [comment]
Thanks @Skrypt. My hope was to not take on responsibility for stuff Orchard is already doing just fine. I want TheAdmin styles to keep working as they always have and just add mine after them. If I understand your suggestion, I'd be *replacing* TheAdmin styles with a copy in my theme and then any changes that the built-in TheAdmin theme gets will not be see in my app (unless I monitor that file for changes and then merge those changes into my copy).

What I want would be the equivalent of post proccessing the HTML just before it goes to the client, finding the link tag to TheAdmin's .css, and inserting my own link to my .css file on the next line. That way I'll automatically get updates to the TheAdmin's css. Obviously actually doing it that way isn't ideal but that is the effective result I want. Any options for getting this result?

**Reply by Skrypt** [reply]
Then you can create a custom CSS declared right after ours that will override TheAdmin.css styles. If you are using Nuget packages there is no way to reuse the .scss files that are included in the /Assets folder of TheAdmin theme. But that won't prevent from getting breaking changes eventually either if you do so.

**Reply by fuzl-llc** [reply]
> create a custom CSS declared right after ours that will override TheAdmin.css styles

Doesn't this still mean in my theme I have to create a copy of TheAdmin's Layout file that declares TheAdmin.css styles so I can declare my own style right after it? I've done that and it works but now I'm responsible for the copy/pasted Layout file and its evolution over time when I just want a tiny change/addition to it.

**Comment by lampersky** [comment]
@fuzl-llc there is also a way to override the whole **css** file content, let say you are using **TheAdmin** theme, go to your web app, and inside **wwwroot** directory create **css** directory with **TheAdmin.css** file with proper content:

![image](https://user-images.githubusercontent.com/6403130/194125081-6a21f72d-757a-4aae-ac86-bdb9ca67c3d6.png)

as you can see on above screenshot, you can do the same with js files

this way, you can append whatever you want

**Reply by fuzl-llc** [reply]
Thanks @lampersky, but again, wouldn't this require me to copy/paste the whole existing css file into my project and therefore take responsibility for updating it (e.g. if the Orchard team makes a change to it, I'd have to notice that change and merge those changes with my file)? I want TheAdmin theme's Layout and css files to keep being used. I want to add my styles on top of those to override/change as needed but still get any updates the Orchard team makes.

**Comment by lampersky** [comment, marked as answer]
@fuzl-llc so if the case is to just add your custom **css** and **js** on top of an existing admin theme you can do something like this:

![image](https://user-images.githubusercontent.com/6403130/194315908-0ccfee85-e5de-469e-be1e-03a3de784f00.png)

so assuming that latest admin theme is **1.0.0** and latest blog theme is **6.0.7**, the idea is to define resources with same name as existing ones, but with highest possible version number (let's say fake), and add dependency to the real existing version, this way orchard will inject original scripts/styles and then your custom :)

here it is an example:

![image](https://user-images.githubusercontent.com/6403130/194317007-e7cc24e9-2aa7-4495-ba8f-bbafd99bd9b4.png)

![image](https://user-images.githubusercontent.com/6403130/194316421-2011f226-5478-40cb-b744-e7fb17fff0f5.png)

**Reply by Skrypt** [reply]
I like that solution @lampersky

**Reply by fuzl-llc** [reply]
Excellent! It works.

The only thing I had to change was to use the full path to where the CSS was in my theme. Also, it seems the "highest possible version number" doesn't really work. I tried 100.0.0 and it doesn't work. I could only change the last revision number and have it work (unless I changed the version dependency too, e.g. from 1.0.0 to 1.0).

This gets 99% of what I wanted and will use it, thanks!

The 1% it doesn't have is that it could still break when/if Orchard upgrades the version number above 1.0.0. I *do not* have to copy/paste CSS or maintain changes within the file - I just have to notice if that version changes so that is a huge improvement. It seems there may be some slightly better numbers to use though, e.g. I tried `.SetDependencies("admin:1.0")` and `.SetVersion("1.100.0");` and it works which I'm guessing means I won't have to do anything if the Orchard team updates that third/revision number and only if they update major or minor. I couldn't get it to work for modifying the other numbers but this is good enough for me for now.  Thanks again!

**Reply by thecodecat23** [reply]
Hi there,

I came across your post while searching for a way to customize the base theme (TheBlogTheme). I just want to change the colors defined in the variables.less file. I attempted the solution provided by creating a new site.css file and configuring the manifest in the ResourceManagementOptionsConfiguration class, however, the method is not being called. I also tried creating a Startup.cs and registering the ResourceManagementOptionsConfiguration, but even the startup method is not being called. Can you please help me understand what I'm doing wrong?"

UPDATE: After successfully implementing the Startup class and the ResourceManagementOptionsConfiguration, I found that the issue is with the path I was setting for the css file I wanted to substitute. If anyone has had a similar issue or has any advice on how to resolve this, I would greatly appreciate your input. Thank you.
